### PR TITLE
Create elastic recoil xs base uo

### PR DESCRIPTION
**Pull request by snschune**

### CODE REVIEW

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  virtual ~ElasticRecoilCrossSectionUserObject();
+
+  virtual void initialize() override;
+
+  virtual void execute() override;
+
+  virtual void finalize() override;
+
+protected:
+
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: no new line after `protected`

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  /// Function representing the reactor neutron spectrum.
+  Function & _neutron_spectrum;
+
+  /// Function representing the epithermal spectrum scattering law.
+  Function & _scattering_law;
+
+  /// Function representing the simple elastic recoil cross section calculation.
+  Function & _simple_erxs;
+
+  Real legendreP(unsigned int  n, Real x);
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: Declare methods first, then all variables. Is there an extra space before `n`?

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+protected:
+
+  /// Function representing the reactor neutron spectrum.
+  Function & _neutron_spectrum;
+
+  /// Function representing the epithermal spectrum scattering law.
+  Function & _scattering_law;
+
+  /// Function representing the simple elastic recoil cross section calculation.
+  Function & _simple_erxs;
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: I think this is just the `elastic_scattering_xs`?

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+  Real legendreP(unsigned int  n, Real x);
+
+  QBase * _quadrature;
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: Add doxygen comments here. You can have one comment applying to all three lines using:

```
///@{
Real a;
Real b;
///@}
```

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  std::vector<Real> _quad_points;
+  std::vector<Real> _quad_weights;
+
+  std::vector<Real> _atomic_mass;
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: `const`

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  std::vector<unsigned int> _isotope_type;
+  unsigned int _number_isotope;
+  std::vector<Real> _neutron_energy_limits;
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: const

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  std::vector<Real> _recoil_energy_limits;
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: const

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+  Real _L;
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: const

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  std::vector<Real> _xi_g;
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: doxygen comment

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  Real _gamma;
+  unsigned int findNeutronEnergyGroup(Real energy);
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: `unsigned int findNeutronEnergyGroup(Real energy) const;`

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+  std::vector<std::vector<std::vector<Real>>> _elastic_recoil_xs;
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: state that this is the quantity we compute in a doxygen comment

File: src/base/MagliteApp.C [hunk 12]
```diff
@@ -4,6 +4,12 @@
 #include "ModulesApp.h"
 #include "MooseSyntax.h"
 
+
+/*
+ * User Objects
```
Comment (snschune) [inline src/base/MagliteApp.C]: Remove these three lines. The comment has no information content.

File: src/base/MagliteApp.C [hunk 13]
```diff
@@ -46,6 +52,8 @@ MagliteApp__registerObjects(Factory & factory)
 void
 MagliteApp::registerObjects(Factory & factory)
 {
+  registerUserObject(ElasticRecoilCrossSectionUserObject);
+
```
Comment (snschune) [inline src/base/MagliteApp.C]: unnecessary line

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+  const std::vector<Point> & temp_points = _quadrature->get_points();
+  for (auto & pt : temp_points)
+  {
+    _quad_points.push_back(pt(0));
+    _console << pt(0) << std::endl;
+  }
+  _quad_weights = _quadrature->get_weights();
+
+
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: extra lines.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+Real
+ElasticRecoilCrossSectionUserObject::legendreP(unsigned int  n, Real x)
+{
+  switch (n)
+  {
+    case 0:
+      return 1;
+
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: I'd remove the blank lines here.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    case 5:
+      return 0.125 * (63 * pow(x, 5) - 70 * pow(x, 3) + 15 * x);
+
+    default:
+      mooseError("Implementation of Legendre polynomials goes up to n = 5");
+  }
+}
+
+/// Class destructor
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: obvious comment, can be removed

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+ElasticRecoilCrossSectionUserObject::~ElasticRecoilCrossSectionUserObject()
+{
+  delete _quadrature;
+}
+
+void
+ElasticRecoilCrossSectionUserObject::initialize()
+{
+  /// Loop to calculate neutron spectrum over group g xi(E)
+  unsigned int G = _neutron_energy_limits.size() - 1;
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: if `_L` is a member variable, make `_G` and `_T` members, too.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+  _xi_g.resize(G);
+  for (unsigned int g = 0; g < G; ++g)
+  {
+    Real E_max = _neutron_energy_limits[g];
+    Real E_min = _neutron_energy_limits[g + 1];
+
+    unsigned int nqp = _quad_points.size();
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: `_nqp` as global variable?

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+
+    for (unsigned int p = 0; p < nqp; ++p)
+    {
+      Real E = 0.5 * (E_max - E_min) * _quad_points[p] + 0.5 * (E_max + E_min);
+      Real w_E = 0.5 * _quad_weights[p] * (E_max - E_min);
+
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: remove empty line.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+/// Find the neutron energy group
+unsigned int
+ElasticRecoilCrossSectionUserObject::findNeutronEnergyGroup(Real energy)
+{
+  unsigned int G = _neutron_energy_limits.size() - 1;
+  Real E_max = _neutron_energy_limits[0];
+  unsigned int group;
+
+  for (unsigned int g = 0; g < G; ++g)
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: I don't think this is implemented correctly. 

If `energy > _neutron_energy_limits[0]` should error out. Same for the case `energy < _neutron_energy_limits[-1]`.

Comment (snschune) [reply src/userobject/ElasticRecoilCrossSectionUserObject.C]:
```
if (energy > _neutron_energy_limits[0] || energy < _neutron_energy_limits.back())
  mooseError("A meaningful message");

for (unsigned int g = 0; g < G; ++g)
  if (energy > _neutron_energy_limits[g + 1])
    return g;
```

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+      group = G;
+    }
+  }
+  return group;
+}
+
+void
+ElasticRecoilCrossSectionUserObject::execute()
+{
+  _gamma = 4 * _atomic_mass[0] / std::pow(( _atomic_mass[0] + 1),2);
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: `_gamma`  does not change except when changing `_atomic_mass`. I think we want one UserObject for 
each nuclide. Hence `_atomic_mass` should be a single Real number. Then `_gamma` can be computed in the constructor.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+    _elastic_recoil_xs.resize(_L);
+    for (unsigned int l = 0; l < _L; ++l)
+    {
+      _elastic_recoil_xs[l].resize(T);
+      for (unsigned int t = 0; t < T; ++t)
+        _elastic_recoil_xs[l][t].resize(G);
+    }
+
+    // For all T bins
+    // there should be a loop over l here but for now I set l = 0
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: comment addressed so remove it

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+            Real w_theta = _quad_weights[i_theta] * libMesh::pi;
+
+            Real Ei = 2 * T_i / (_gamma * (1 - std::cos(theta_c)));
+
+            unsigned int g = findNeutronEnergyGroup(Ei);
+
+            _elastic_recoil_xs[l][t][g] += 0.5 * std::cos(0.5 * theta_c) * _scattering_law.value(Ei, Point()) *
+                                          legendreP(l, std::sin(0.5 * theta_c)) * w_T * w_theta / _xi_g[g];
+
+            _console << _elastic_recoil_xs[l][t][g] << std::endl;
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: fine for testing, don't forget to remove.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+
+template <>
+InputParameters
+validParams<ElasticRecoilCrossSectionUserObject>()
+{
+  InputParameters params = validParams<GeneralUserObject>();
+  params.addRequiredParam<std::vector<Real>>(
+      "atomic_mass", "Atomic Mass of the isotopes");
+  params.addRequiredParam<std::vector<unsigned int>>(
+      "isotope_type", "Type of the target isotopes");
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: make this a simple Real; a UserObject deals with only a single nuclide.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 25]
```diff
@@ -202,4 +206,23 @@ ElasticRecoilCrossSectionUserObject::execute()
 void
 ElasticRecoilCrossSectionUserObject::finalize()
 {
+  std::ofstream output_file;
+  output_file.open ("erxs_output.csv");
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: extra space after open

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 26]
```diff
@@ -202,4 +206,23 @@ ElasticRecoilCrossSectionUserObject::execute()
...
+
+  unsigned int T = _recoil_energy_limits.size() - 1;
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: make these global. also there is an extra space in front..not that is matters.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 27]
```diff
@@ -202,4 +206,23 @@ ElasticRecoilCrossSectionUserObject::execute()
...
+  unsigned int G = _neutron_energy_limits.size() - 1;
+
+for (unsigned int l = 0; l < _L; ++l)
+{
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: loops that have a single statement don't need `{`
e.g.
```
for (unsigned int j = 0; j < J; ++j)
  for (unsigned int i = 0; i < I; ++i)
    if (i == j)
      _console << "i  is equal to j" << std::endl;
```

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 28]
```diff
@@ -202,4 +206,23 @@ ElasticRecoilCrossSectionUserObject::execute()
...
+  {
+    for (unsigned int t = 0; t < T; ++t) // t columns
+    {
+      output_file << _elastic_recoil_xs[0][t][g] << ',';
+    }
+      output_file << std::endl; // l matrices
+  }
+  output_file << std::endl << std::endl;
+}
+  output_file.close();
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: why is this outside the last `}`

File: include/userobject/ElasticRecoilCrossSectionUserObject.h (hunk 3 above)
Comment (marinasessim) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h]: _simple_erxs is the function that calls the variable elastic_scattering_xs. Should I rename them?

Comment (marinasessim) [reply include/userobject/ElasticRecoilCrossSectionUserObject.h]: _simple_erxs is the parsed function that has the expression for the cross section and _elastic_recoil_xs is the variable for the calculations ([l][t][g])

File: include/userobject/ElasticRecoilCrossSectionUserObject.h [hunk 29]
```diff
@@ -0,0 +1,93 @@
...
+template <>
+InputParameters validParams<ElasticRecoilCrossSectionUserObject>();
+
+class ElasticRecoilCrossSectionUserObject : public GeneralUserObject
+{
+public:
+  ElasticRecoilCrossSectionUserObject(const InputParameters & parameters);
+
+  /**
+   * Class destructor
```
Comment (snschune) [inline include/userobject/ElasticRecoilCrossSectionUserObject.h:23]: remive this comment, it's obvious that it's a destructor.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 30]
```diff
@@ -0,0 +1,202 @@
...
+    _neutron_energy_limits(getParam<std::vector<Real>>("neutron_energy_limits")),
+    _recoil_energy_limits(getParam<std::vector<Real>>("recoil_energy_limits"))
+{
+   /// See "enum_order.h and enum_quadrature_type.h"
+  _quadrature = QBase::build(QGAUSS, 1, FORTYTHIRD).release();
+
+  /// Extract 1 dim quadrature rule from libmesh and store in points and weights
+  const std::vector<Point> & temp_points = _quadrature->get_points();
+  for (auto & pt : temp_points)
+  {
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C:55]: the 
```
{
}
````
are not necessary

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 31]
```diff
@@ -0,0 +1,202 @@
...
+
+Real
+ElasticRecoilCrossSectionUserObject::legendreP(unsigned int  n, Real x)
+{
+  switch (n)
+  {
+    case 0: return 1;
+    case 1: return x;
+    case 2: return 0.5 * (3 * pow(x, 2) - 1);
+    case 3: return  0.5 * (5 * pow(x, 3) - 3 * x);
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C:73]: extra space.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 32]
```diff
@@ -0,0 +1,202 @@
...
+    }
+  }
+}
+
+/// Find the neutron energy group given a neutron energy
+unsigned int
+ElasticRecoilCrossSectionUserObject::findNeutronEnergyGroup(Real energy)
+{
+  for (unsigned int g = 0; g < _G; ++g)
+  {
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C:111]: the `{` us unnecessary here.
```
for (blabla)
  if (more blalbla)
    do someting;
```
is fine!

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 33]
```diff
@@ -0,0 +1,202 @@
...
+
+      _xi_g[g] += w_E * _neutron_spectrum.value(E, Point());
+    }
+  }
+}
+
+/// Find the neutron energy group given a neutron energy
+unsigned int
+ElasticRecoilCrossSectionUserObject::findNeutronEnergyGroup(Real energy)
+{
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C:109]: add mooseAsserts here to make sure `energy` is within bounds.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 34]
```diff
@@ -0,0 +1,202 @@
...
+  }
+  /// C++ is stupid so this is necessary
+  mooseError("Should never get here");
+  return 0;
+}
+
+void
+ElasticRecoilCrossSectionUserObject::execute()
+{
+  /// Size the cross section array
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C:123]: indentation.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 35]
```diff
@@ -0,0 +1,202 @@
...
+    _erxs_coeff.resize(_L);
+    for (unsigned int l = 0; l < _L; ++l)
+    {
+      _erxs_coeff[l].resize(_T);
+      for (unsigned int t = 0; t < _T; ++t)
+      {
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C]: this pair of brackets is unnecessary.

File: src/userobject/ElasticRecoilCrossSectionUserObject.C [hunk 36]
```diff
@@ -0,0 +1,202 @@
...
+  std::ofstream output_file;
+  output_file.open ("erxs_output.csv");
+  for (unsigned int l = 0; l < _L; ++l)
+  {
+    for (unsigned int g = 0; g < _G; ++g)
+    {
+      for (unsigned int t = 0; t < _T; ++t)
+      {
+        if (t < _T - 1)
+        output_file << _erxs_coeff[l][t][g] << ',';
```
Comment (snschune) [inline src/userobject/ElasticRecoilCrossSectionUserObject.C:203]: indentation
```
if (condition)
  I am indented by 2 spaces;
```

**Review by snschune — COMMENTED** [review]
Very few formatting issues, otherwise it's good. 
Let's see if the results make sense and move on!

**Comment by snschune** [comment]
Little bit of formatting and the we'll need to change output format. I'll merge this now because this PR has become intractable.

**Pull request merged**